### ISSUE TITLE
Changes in the results

### ISSUE THREAD

**Issue by AlsalehAbdullah** [aub-mind/arabert#121]
I have used the AraBERT v2 and v0.2 for my classification experiments. The results produced in February is different in June although I used the same dataset and training process. I sat the manual seed and random seed so that I can experiment again knowing that the training is the same process. 
```
seed_val = 12

random.seed(seed_val)
np.random.seed(seed_val)
torch.manual_seed(seed_val)
torch.cuda.manual_seed_all(seed_val)
```

So, was there any updates to the model between February and June that caused different outcome in my experiments? Or is it something else? 

Thank you
Abdullah

**Comment by WissamAntoun** [comment]
The model weights have not been changed at all. I only changed the preprocessor to deal with some issues like the missing "/" token.
You can try and clone the repository from a commit made in February and test it again.

Also I'm not sure if this is related https://tanelp.github.io/posts/a-bug-that-plagues-thousands-of-open-source-ml-projects/

**Comment by salama4ai** [comment]
if you change the version of python or the pytorch version or other package, this might change the output of `torch.manual_seed(seed_val)
`and`torch.cuda.manual_seed_all(seed_val)` for the same seed_val

**Comment by AlsalehAbdullah** [comment]
@salama4ai - I don't know which Python or Torch version when I was experimenting in February, all of my experiments are being done using Google Colab so I'm not aware if there was any version updates. 

@WissamAntoun - I'll have a look at the article, it looks related. I'm going to clone the repository in February in my local machine, but is there a way to do it using Google Colab? 

Thank you

**Comment by WissamAntoun** [comment]
You click on the github history, then copy the hash of the commit that you want to go back to. then clone it as follows

```
git clone https://github.com/aub-mind/arabert/
cd arabert && git checkout 6a58ca118911ef311cbe8cdcdcc1d03601123291 <-- replace this with the hash you want
```

**Comment by AlsalehAbdullah** [comment]
@WissamAntoun 

I got the hash of the commit (23 Jan) and I applied it to the Google Colab notebook. Then, I've done the experiment three times for the same training, valid and testing sets and the model that I used was _BertForSequenceClassification_ and here are the results of AraBERT v0.2
The research I'm doing is to feed the network with pairs of input sequences to predict if they are related or not. Label '1' means they are related and '0' not related. Here's the paper https://aclanthology.org/2021.wanlp-1.19.pdf

**First results:**

Confusion Matrix:  
 [[755  44]
 [138 745]]


Accuracy Score:  0.89179548156956


Classification Report:  
               precision    recall  f1-score   support

           0       0.85      0.94      0.89       799
           1       0.94      0.84      0.89       883

    accuracy                           0.89      1682
   macro avg       0.89      0.89      0.89      1682
weighted avg       0.90      0.89      0.89      1682


**Second results:**

Confusion Matrix:  
 [[754  45]
 [100 783]]


Accuracy Score:  0.9137931034482759


Classification Report:  
               precision    recall  f1-score   support

           0       0.88      0.94      0.91       799
           1       0.95      0.89      0.92       883

    accuracy                           0.91      1682
   macro avg       0.91      0.92      0.91      1682
weighted avg       0.92      0.91      0.91      1682


**Third results:**

Confusion Matrix:  
 [[751  48]
 [118 765]]


Accuracy Score:  0.901307966706302


Classification Report:  
               precision    recall  f1-score   support

           0       0.86      0.94      0.90       799
           1       0.94      0.87      0.90       883

    accuracy                           0.90      1682
   macro avg       0.90      0.90      0.90      1682
weighted avg       0.90      0.90      0.90      1682



What do you think? 

Thank you 
Abdullah

**Comment by WissamAntoun** [comment]
The results look on par with paper.

Can I ask how are you loading the data into the model? Are you adding a `[SEP]` token between the 2 sequences?

**Comment by AlsalehAbdullah** [comment]
I used the `batch_encode_plus` for tokenizing the pair of sequences.

```
tokenized_train=tokenizer.batch_encode_plus(
    df_train[["Verse1", "Verse2"]].values.tolist(),
    max_length=128,
    pad_to_max_length=True,
    truncation=True,
    return_tensors='pt')
```

here is the output of `tokenized_train['input_ids'][0]`:
```
tensor([    2,  9448, 14223,   323,   860, 53032,   319,  1571, 20315, 28289,
          305,  6327, 11229,  1946,  2705,   111,   209,  1493, 34064, 11763,
            3, 16267,   141, 14071, 33948, 58521,  2209, 40398,  1199,  3120,
        31734,  4402, 42752, 30861,   319,     3,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0])
```

and we convert them to tokens, `tokenizer.convert_ids_to_tokens(tokenized_train['input_ids'][1])`

```
['[CLS]',
 'انما',
 'السبيل',
 'على',
 'الذين',
 'يظلم',
 '##ون',
 'الناس',
 'ويب',
 '##غون',
 'في',
 'الارض',
 'بغير',
 'الحق',
 'اول',
 'ئ',
 '##ك',
 'لهم',
 'عذاب',
 'اليم',
 '[SEP]',
 'وقالوا',
 'ي',
 'ايه',
 'الساحر',
 'ادع',
 'لنا',
 'ربك',
 'بما',
 'عهد',
 'عندك',
 'اننا',
 'لمه',
 '##تد',
 '##ون',
 '[SEP]',
 '[PAD]',
 .
 '[PAD]',
 '[PAD]']
```

**Comment by WissamAntoun** [comment]
Great!

**Comment by AlsalehAbdullah** [comment]
So is it normal producing different results in neural networks? I know they weren't dramatic changes but I just would like to know moving forward from now.

**Comment by WissamAntoun** [comment]
Well, ideally it should produce the same result if all seeds were set correctly, and you are using the exact same hardware.
My guess is that there is a seed changing somewhere (https://pytorch.org/docs/stable/notes/randomness.html).

The best thing is to average multiple runs, to guarantee statistical significance.

**Comment by AlsalehAbdullah** [comment]
I'm trying to figure it out if I sat it up correctly :D, I'll have a look at the link. Thank you!

Sorry one question and I've probably should've mentioned before, do you think that RandomSampler when included in the DataLoader could potentially change the training data selection even when the seed were placed in the training?

```
train_data=TensorDataset(tokenized_train['input_ids'],
                         tokenized_train['attention_mask'],
                         tokenized_train['token_type_ids'],
                         labels_train)

train_sampler=RandomSampler(train_data)

train_dataloader=DataLoader(train_data,
                            sampler=train_sampler,
                            batch_size=batch_size)
```

**Comment by WissamAntoun** [comment]
AH I think that's it. Check the last section in the link I sent on how to seed the data loader in pytorch

**Comment by AlsalehAbdullah** [comment]
Oh ok, I will check it and try it.

Thank you @WissamAntoun for your time and feedback
